cpp/src/arrow/type_traits.h

```diff
@@ -389,193 +389,258 @@ using void_t = typename make_void<Ts...>::type;
 
 template <typename T, typename R = void>
-using enable_if_base_list =
-    typename std::enable_if<std::is_base_of<BaseListType, T>::value, R>::type;
+using enable_if_date = enable_if_t<is_date_type<T>::value, R>;
+
+template <typename T>
+using is_time_type = std::is_base_of<TimeType, T>;
 
 template <typename T, typename R = void>
-using enable_if_list =
-    typename std::enable_if<std::is_base_of<ListType, T>::value, R>::type;
+using enable_if_time = enable_if_t<is_time_type<T>::value, R>;
+
+template <typename T>
+using is_timestamp_type = std::is_base_of<TimestampType, T>;
 
 template <typename T, typename R = void>
-using enable_if_large_list =
-    typename std::enable_if<std::is_base_of<LargeListType, T>::value, R>::type;
+using enable_if_timestamp = enable_if_t<is_timestamp_type<T>::value, R>;
+
+template <typename T>
+using is_duration_type = std::is_base_of<DurationType, T>;
 
 template <typename T, typename R = void>
-using enable_if_fixed_size_list =
-    typename std::enable_if<std::is_base_of<FixedSizeListType, T>::value, R>::type;
+using enable_if_duration = enable_if_t<is_duration_type<T>::value, R>;
+
+// Attribute differentiation
+
+template <typename T>
+using is_primitive_ctype = std::is_base_of<PrimitiveCType, T>;
 
 template <typename T, typename R = void>
-using enable_if_number = typename std::enable_if<is_number_type<T>::value, R>::type;
+using enable_if_primitive_ctype = enable_if_t<is_primitive_ctype<T>::value, R>;
 
-namespace internal {
+template <typename T>
+using has_c_type = std::integral_constant<bool, is_primitive_ctype<T>::value ||
+                                                    is_temporal_type<T>::value>;
 
-// The partial specialization will match if T has the ATTR_NAME member
-#define GET_ATTR(ATTR_NAME, DEFAULT)                             \
-  template <typename T, typename Enable = void>                  \
-  struct GetAttr_##ATTR_NAME {                                   \
-    using type = DEFAULT;                                        \
-  };                                                             \
-                                                                 \
-  template <typename T>                                          \
-  struct GetAttr_##ATTR_NAME<T, void_t<typename T::ATTR_NAME>> { \
-    using type = typename T::ATTR_NAME;                          \
-  };
+template <typename T, typename R = void>
+using enable_if_has_c_type = enable_if_t<has_c_type<T>::value, R>;
 
-GET_ATTR(c_type, void)
-GET_ATTR(TypeClass, void)
+template <typename T>
+using has_string_view = std::integral_constant<bool, is_binary_like_type<T>::value ||
+                                                         is_string_like_type<T>::value>;
 
-#undef GET_ATTR
+template <typename T, typename R = void>
+using enable_if_has_string_view = enable_if_t<has_string_view<T>::value, R>;
 
-}  // namespace internal
+template <typename T>
+using has_no_extra_meta = std::is_base_of<NoExtraMeta, T>;
```

Comment:
What does "no extra meta" mean? If we're refactoring, we might as well give this a readable name (for example "is_parameterless"?).

Reply:
I'm not so sure, the place I've seen it referenced is https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/json_internal.cc#L132-L375, which is used for ipc serialization.

We can probably remove this traits and use explicit visitor methods where required.

Reply:
Removed.